jba/src/jba/tests/tests_parser.py

```diff
@@ -0,0 +1,246 @@
+class ExceptionData:
+    path: str
+    line_number: int
+    column_number: int
+    message: str
+
+
+def parse_gradle_logs(submission: pd.Series, gradle_logs_path: Path) -> Tuple[Optional[str], Optional[str]]:
+    """
+    Parse Gradle stderr and test logs.
+
+    :param submission: User's submission data.
+    :param gradle_logs_path: Path to a folder with gradle logs.
+    :return: Parsed stderr and test logs as json strings.
+    """
+    submission_id = submission[EduColumnName.ID.value]
+
+    logger.info(f'Parsing submission#{submission_id}')
+
+    submission_gradle_logs_path = gradle_logs_path / str(submission_id)
+    if not submission_gradle_logs_path.exists():
+        logger.warning("Submission Gradle logs are missing")
+        return None, None
+
+    section_name = submission.get(EduColumnName.SECTION_NAME.value)
+    lesson_name = submission[EduColumnName.LESSON_NAME.value]
+    task_name = submission[EduColumnName.TASK_NAME.value]
+
+    task_path = Path('' if section_name is None else section_name) / lesson_name / task_name
+
+    parsed_stderr_logs = _parse_stderr_logs(str(task_path), submission_gradle_logs_path)
+    if parsed_stderr_logs:
+        parsed_stderr_logs = json.dumps([asdict(exception) for exception in parsed_stderr_logs])
+
+    parsed_test_logs = _parse_test_logs(submission_gradle_logs_path)
+    if parsed_test_logs:
+        parsed_test_logs = json.dumps([asdict(test) for test in parsed_test_logs])
+
+    return parsed_stderr_logs, parsed_test_logs
+
+
+def _parse_stderr_logs(task_path: str, submission_gradle_logs_path: Path) -> Optional[List[ExceptionData]]:
+    """
+    Parse Gradle stderr logs into list of `ExceptionData`.
+
+    :param task_path: Relative path from a course root to a task.
+    :param submission_gradle_logs_path: Path to a file with gradle stderr logs.
+    :return: List of `ExceptionData`.
+    """
+    logger.info('Parsing stderr logs')
+
+    stderr_logs_file = submission_gradle_logs_path / GRADLE_STDERR_LOGS_FILE
+    if not stderr_logs_file.exists():
+        logger.info('Stderr logs file does not exist')
+        return None
+
+    exceptions = []
+    with open(stderr_logs_file) as file:
+        for line in file:
+            match = re.match(EXCEPTION_REGEXP, line)
+            if not match:
+                continue
+
+            full_file_path = match.group(1)
+            relative_file_path = full_file_path[full_file_path.find(task_path) + len(task_path) + 1:]
+
+            exceptions.append(
+                ExceptionData(
+                    path=relative_file_path,
+                    line_number=int(match.group(2)),
+                    column_number=int(match.group(3)),
+                    message=match.group(4),
+                ),
+            )
+
+    if not exceptions:
+        logger.info('There are no exceptions in the stderr logs file')
+
+    return exceptions
+
+
+def _parse_test_logs(submission_gradle_logs_path: Path) -> Optional[List[TestData]]:
```

Comment:
I'w prefer to separate this function into several ones, also maybe it is better to move HTML parsing into another file